pkg/asset/cluster/metadata.go

```diff
@@ -0,0 +1,115 @@
+package cluster
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/aws"
+	"github.com/openshift/installer/pkg/asset/cluster/libvirt"
+	"github.com/openshift/installer/pkg/asset/cluster/openstack"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/types"
+	"github.com/pkg/errors"
+)
+
+const (
+	metadataFileName = "metadata.json"
+)
+
+// Metadata contains information needed to destroy clusters.
+type Metadata struct {
+	file *asset.File
+}
+
+var _ asset.WritableAsset = (*Metadata)(nil)
+
+// Name returns the human-friendly name of the asset.
+func (m *Metadata) Name() string {
+	return "Metadata"
+}
+
+// Dependencies returns the direct dependencies for the metadata
+// asset.
+func (m *Metadata) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&installconfig.ClusterID{},
+		&installconfig.InstallConfig{},
+	}
+}
+
+// Generate generates the metadata asset.
+func (m *Metadata) Generate(parents asset.Parents) (err error) {
+	clusterID := &installconfig.ClusterID{}
+	installConfig := &installconfig.InstallConfig{}
+	parents.Get(clusterID, installConfig)
+
+	if installConfig.Config.Platform.None != nil {
+		return errors.New("metadata cannot be created with platform set to 'none'")
+	}
+
+	metadata := &types.ClusterMetadata{
+		ClusterName: installConfig.Config.ObjectMeta.Name,
+		ClusterID:   clusterID.ClusterID,
+	}
+
+	switch {
+	case installConfig.Config.Platform.AWS != nil:
+		metadata.ClusterPlatformMetadata.AWS = aws.Metadata(clusterID.ClusterID, installConfig.Config)
+	case installConfig.Config.Platform.Libvirt != nil:
+		metadata.ClusterPlatformMetadata.Libvirt = libvirt.Metadata(installConfig.Config)
+	case installConfig.Config.Platform.OpenStack != nil:
+		metadata.ClusterPlatformMetadata.OpenStack = openstack.Metadata(clusterID.ClusterID, installConfig.Config)
+	default:
+		return errors.Errorf("no known platform")
+	}
+
+	data, err := json.Marshal(metadata)
+	if err != nil {
+		return errors.Wrap(err, "failed to Marshal ClusterMetadata")
+	}
+
+	m.file = &asset.File{
+		Filename: metadataFileName,
+		Data:     data,
+	}
+
+	return nil
+}
+
+// Files returns the FileList generated by the asset.
+func (m *Metadata) Files() []*asset.File {
+	return []*asset.File{m.file}
+}
+
+// Load returns error if the metadata is already on-disk, because we want to
+// prevent user from accidentally re-launching the cluster.
+func (m *Metadata) Load(f asset.FileFetcher) (found bool, err error) {
+	_, err = f.FetchByName(metadataFileName)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, errors.Errorf("%q already exists.  There may already be a running cluster", metadataFileName)
+}
+
+// LoadMetadata loads the cluster metadata from an asset directory.
+func LoadMetadata(dir string) (*types.ClusterMetadata, error) {
```

Comment:
This function is never called.

Reply:
> This function is never called.

Yes it is?

```console
$ git grep LoadMetadata origin/pr/1070
origin/pr/1070:pkg/asset/cluster/metadata.go:// LoadMetadata loads the cluster metadata from an asset directory.
origin/pr/1070:pkg/asset/cluster/metadata.go:func LoadMetadata(dir string) (*types.ClusterMetadata, error) {
origin/pr/1070:pkg/destroy/bootstrap/bootstrap.go:      metadata, err := cluster.LoadMetadata(dir)
origin/pr/1070:pkg/destroy/destroyer.go:        metadata, err := cluster.LoadMetadata(rootDir)
```

It's originally from 3f4fe57414 (#493).

Reply:
Yeah, sorry. Bad grep.